File: shared/types/session.ts
// TS mirror of the backend session/turn schemas (app/schemas/session.py, turn.py).
import type { Difficulty, Role } from "./scenario";

export type SessionStatus = "active" | "completed" | "abandoned";
export type Speaker = "user" | "persona";

export interface TurnOut {
  index: number;
  speaker: Speaker;
  text: string;
  emotion: string | null;
  severity_after: number | null;
  confidence_after: number | null;
  evaluation: Record<string, unknown> | null;
  created_at: string;
}

export interface EscalationEvent {
  event: string;
  severity: number;
  turn_index: number;
  at: string;
}

export interface SessionState {
  severity: number;
  severity_start: number;
  confidence: number;
  objectives_met: string[];
  escalation_timeline: EscalationEvent[];
  status: SessionStatus;
}

export interface SessionCreate {
  scenario_id: string;
  role: Role;
  difficulty: Difficulty;
}

export interface SessionOut {
  id: string;
  scenario_id: string;
  role: Role;
  difficulty: Difficulty;
  status: SessionStatus;
  severity: number;
  severity_start: number;
  confidence: number;
  score: number | null;
  objectives_met: string[];
  escalation_timeline: EscalationEvent[];
  created_at: string;
  ended_at: string | null;
}

export interface SessionDetail extends SessionOut {
  turns: TurnOut[];
}

export interface SessionListItem {
  id: string;
  scenario_id: string;
  role: Role;
  difficulty: Difficulty;
  status: SessionStatus;
  score: number | null;
  created_at: string;
  ended_at: string | null;
}

File: backend/app/schemas/session.py
"""Drill session request/response schemas."""
from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict

from app.models.session import SessionStatus
from app.schemas.common import Difficulty, Role
from app.schemas.turn import TurnOut


class SessionCreate(BaseModel):
    scenario_id: str
    role: Role
    difficulty: Difficulty


class SessionState(BaseModel):
    severity: int
    severity_start: int
    confidence: int
    objectives_met: list[str]
    escalation_timeline: list[dict[str, Any]]
    status: SessionStatus


class SessionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    scenario_id: str
    role: Role
    difficulty: Difficulty
    status: SessionStatus
    severity: int
    severity_start: int
    confidence: int
    score: Optional[float] = None
    objectives_met: list[str]
    escalation_timeline: list[dict[str, Any]]
    created_at: datetime
    ended_at: Optional[datetime] = None


class SessionDetail(SessionOut):
    turns: list[TurnOut] = []


class SessionListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    scenario_id: str
    role: Role
    difficulty: Difficulty
    status: SessionStatus
    score: Optional[float] = None
    created_at: datetime
    ended_at: Optional[datetime] = None
